pkg/didcomm/protocol/introduce/persistence.go

```diff
@@ -0,0 +1,77 @@
+/*
+Copyright SecureKey Technologies Inc. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package introduce
+
+import (
+	"encoding/json"
+
+	"github.com/hyperledger/aries-framework-go/pkg/didcomm/common/service"
+	"github.com/hyperledger/aries-framework-go/pkg/didcomm/protocol/didexchange"
+	"github.com/hyperledger/aries-framework-go/pkg/storage"
+)
+
+const recordPrefix = Introduce + "_record_"
+
+// Persistence is a repository to store data for introduce protocol
+type Persistence struct {
+	store storage.Store
+}
+
+// Record keeps transited data for state machine
+type Record struct {
+	Invitation   *didexchange.Invitation `json:"invitation,omitempty"`
+	Destinations []*service.Destination  `json:"destinations,omitempty"`
+}
+
+// NewPersistence creates repository instance
+func NewPersistence(store storage.Store) *Persistence {
+	return &Persistence{store: store}
+}
+
+// GetRecord returns data by given threadID
+// this function should be used in introduce service
+// state machine behavior depends on this record's data
+func (p *Persistence) GetRecord(thID string) (*Record, error) {
+	var rec *Record
+	if err := p.getData(recordKey(thID), &rec); err != nil {
+		return nil, err
+	}
+
+	return rec, nil
+}
+
+// SaveRecord saves data
+// this function usually is executed by the introduce client
+// NOTE: if the record already exists in the store it will rewrite it
+func (p *Persistence) SaveRecord(thID string, rec *Record) error {
```

Comment:
Ideally client shouldn't save protocol related records and it should be done through service pkg (knowing of lower layer primitives like threadID). What's the use case here ?

Reply:
the client has a bunch of data needed for service, we have similar logic in `didexchange` client e.g `connectionStore: didexchange.NewConnectionRecorder(transientStore, store),`

Reply:
Yes, but not save record - https://github.com/hyperledger/aries-framework-go/blob/master/pkg/didcomm/protocol/didexchange/persistence.go#L195

Reply:
`Record` it is just a name :) which keeps dependency needed for the state machine https://github.com/hyperledger/aries-framework-go/pull/711/files#diff-2be7cff09231a8b321a1f1aed45a4913R24-R28

Reply:
@fqutishat @llorllale @rolsonquadras @sandrask

Reply:
Is the issue that the method should be named to something more specific?
@soluchok @rolsonquadras

Reply:
@troyronda My comment was about the 2nd item that you pointed out. ie - service should be persisting thread related data. ie. un-export the function.

I see the db operations from different layers as follows.
1. Service layer - handle db operations for did-comm message related data. ie, based on threadID or messageID etc.
2. Client layer - handle db operation for agent/application specific data. ie, based on connectionID(which is specific to the agent)

Note - service layer can handle pt#2 as well.